instrumentation/servlet/servlet-3.0/library/src/main/java/io/opentelemetry/instrumentation/servlet/v3_0/Servlet3HttpServerTracer.java

```diff
@@ -116,16 +111,23 @@ private static MappingResolver getMappingResolver(
 
   public Context updateContext(
       Context context, Object servletOrFilter, HttpServletRequest request) {
-    Span span = ServerSpan.fromContextOrNull(context);
-    if (span != null && ServletSpanNaming.shouldUpdateServerSpanName(context)) {
-      String spanName = getSpanNameFromPath(servletOrFilter, request);
-      if (spanName != null) {
-        span.updateName(spanName);
-        ServletSpanNaming.setServletUpdatedServerSpanName(context);
+    updateServerSpanName(context, servletOrFilter, request);
+    return updateContext(context, request);
+  }
+
+  private static void updateServerSpanName(
+      Context context, Object servletOrFilter, HttpServletRequest request) {
+    Span serverSpan = ServerSpan.fromContextOrNull(context);
+    if (serverSpan != null) {
+      ServletSpanNaming servletSpanNaming = ServletSpanNaming.from(context);
+      if (servletSpanNaming.shouldServletUpdateServerSpanName()) {
+        String spanName = getSpanNameFromPath(servletOrFilter, request);
+        if (spanName != null) {
+          serverSpan.updateName(spanName);
+          servletSpanNaming.setServletUpdatedServerSpanName();
```

Comment:
This triple-if pattern can be seen in several different instrumentations. WDYT about adding the following method to `ServletSpanNaming`:

```java
void updateServerSpanNameByServlet(Supplier<String> serverSpanName) {
  Span serverSpan = ServerSpan.fromContextOrNull(context);
  if (serverSpan == null || !shouldServletUpdateServerSpanName()) {
    return;
  }
  if (serverSpanName != null) {
    serverSpan.updateName(serverSpanName);
    setServletUpdatedServerSpanName();
  }
}
```

and then just call:

```java
ServletSpanNaming.from(context)
    .updateServerSpanNameByServlet(() -> getSpanNameFromPath(servletOrFilter, request));
```

in the instrumentation code?

Reply:
I think it is a good idea because every framework will need to make similar checks. Perhaps we could add a `kind` or `priority` to represent that controller can overwrite name from servlet but not vice versa?

Reply:
Well, I thought of copy-pasting this method with s/Servlet/Controller/, but a kind/priority would be fine too.